src/executors.jl

```diff
@@ -49,137 +49,138 @@ and should be retried (and `false` otherwise). The default `retry_on` method ret
 
 # Returns
 
-* `Vector{DispatchResult}`: an array containing a `DispatchResult` for each node in `nodes`,
-  in that order.
+* `Vector{DispatchResult}`: an array containing a `DispatchResult` for each node in
+  `output_nodes`, in that order.
 
 # Throws
 
-* `ExecutorError`: if the context's graph contains a cycle
+* `ExecutorError`: if the constructed graph contains a cycle
 * `CompositeException`/[`DependencyError`](@ref): see documentation for
-  [`dispatch!(::Executor, ::DispatchContext)`](@ref)
+  [`dispatch!(::Executor, ::DispatchGraph)`](@ref)
 """
 function run!{T<:DispatchNode, S<:DispatchNode}(
     exec::Executor,
-    ctx::DispatchContext,
-    nodes::AbstractArray{T},
+    output_nodes::AbstractArray{T},
     input_nodes::AbstractArray{S}=DispatchNode[];
     input_map::Associative=Dict{DispatchNode, Any}(),
     throw_error=true
 )
-    reduced_ctx = copy(ctx)
-    input_node_keys = DispatchNode[
-        n for n in chain(input_nodes, keys(input_map))
-    ]
+    graph = build_graph(output_nodes, Set(chain(input_nodes, keys(input_map))))
 
-    reduced_ctx.graph = subgraph(ctx.graph, nodes, input_node_keys)
-
-    if is_cyclic(reduced_ctx.graph.graph)
+    if is_cyclic(graph.graph)
         throw(ExecutorError(
             "Dispatcher can only run graphs without circular dependencies",
         ))
     end
 
     # replace nodes in input_map with their values
     for (node, val) in chain(zip(input_nodes, imap(fetch, input_nodes)), input_map)
-        node_id = reduced_ctx.graph.nodes[node]
-        reduced_ctx.graph.nodes[node_id] = DataNode(val)
+        node_id = graph.nodes[node]
+        graph.nodes[node_id] = DataNode(val)
     end
 
-    # add_cleanup_nodes!(reduced_ctx; exclude=union(nodes, input_nodes))
-
-    prepare!(exec, reduced_ctx)
-    node_results = dispatch!(exec, reduced_ctx; throw_error=throw_error)
+    prepare!(exec, graph)
+    node_results = dispatch!(exec, graph; throw_error=throw_error)
 
     # select the results requested by the `nodes` argument
-    return DispatchResult[node_results[reduced_ctx.graph.nodes[node]] for node in nodes]
+    return DispatchResult[node_results[graph.nodes[node]] for node in output_nodes]
 end
 
 """
-    add_cleanup_nodes!(ctx::DispatchContext) -> DispatchContext
+    build_graph(output_nodes, input_nodes) -> DispatchGraph
 
-For all nodes with children in a given context, add a `CleanupNode` to the graph which will
-wait for the child nodes to complete, then clean up the node's data.
-
-Return the first argument.
+Construct a graph starting from `input_nodes` and ending in `output_nodes`.
+The graph is created by recursively identifying dependencies of nodes starting with
+`output_nodes` and ending with `input_nodes` (dependencies of `input_nodes` are not added to
+the graph).
 """
-function add_cleanup_nodes!(
-    ctx::DispatchContext;
-    exclude::Vector=DispatchNode[],
+function build_graph{T<:DispatchNode, S<:DispatchNode}(
```

Comment:
Could these just be `DispatchGraph` constructors to minimize cluttering up the namespace?